src/backend/distributed/transaction/transaction_management.c

```diff
@@ -317,12 +322,23 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 			MemoryContext previousContext =
 				MemoryContextSwitchTo(CitusXactCallbackContext);
 
-			if (CurrentCoordinatedTransactionState == COORD_TRANS_PREPARED)
+			if (CurrentCoordinatedTransactionState == COORD_TRANS_PREPARED &&
+				!IsMainDBCommand)
 			{
 				/* handles both already prepared and open transactions */
 				CoordinatedRemoteTransactionsCommit();
 			}
 
+			/*
+			 * If this is a non-Citus main database we should try to commit the prepared
+			 * transactions created by the Citus main database on the worker nodes.
+			 */
+			if (!IsMainDB && MainDBConnection != NULL)
+			{
+				RunCitusMainDBQuery(COMMIT_MANAGEMENT_COMMAND_2PC);
```

Comment:
Invoking `citus_internal.commit_management_command_2pc()` here, in the  `XACT_EVENT_COMMIT` handler makes the server much more susceptible to crashes, because every error thrown here will lead to a process termination. I faced it when [introducing a timeout](https://github.com/citusdata/citus/pull/7286/files#diff-612c5eb96641bee4664f86dfba7817903d4dd5b82c76b6017354d55e01ba3eef) for maintenance connection pool in #7286.
It would be better to signal the maintenance worker to trigger a recovery somehow...
cc @JelteF